GMX-Bots/gmx5-detect-frontruns/src/agent.ts

```diff
@@ -0,0 +1,206 @@
+    const swapEvents = txEvent.filterLog(swapEvent);
+
+    //detect calls to the GMX router
+    if (txEvent.to == router.value) {
+      swapEvents.forEach((swapEvent) => {
+        const {
+          account: accountBack,
+          tokenIn: tokenInBack,
+          tokenOut: tokenOutBack,
+          amountIn: amountInBack,
+          amountOut: amountOutBack,
+        } = swapEvent.args;
+
+        //order tokens in lexicographical order for the map key
+        let token0 = "";
+        let token1 = "";
+        if (tokenInBack > tokenOutBack) {
+          token0 = tokenOutBack;
+          token1 = tokenInBack;
+        } else {
+          token0 = tokenInBack;
+          token1 = tokenOutBack;
+        }
+
+        let sandwichAttack: boolean = false;
+        let victimAddress = "";
+        let frontHash = "";
+        let victimHash = "";
+        let victimTokenIn = "";
+        let victimTokenOut = "";
+        let victimTokenInAmount = 0;
+        let victimTokenOutAmount = 0;
+        let frontrunnerAddress = "";
+        let frontrunnerProfit = 0;
+        let victimAddressAndTokens: string = "";
+        let frontrunnerAddressAndTokens: string = "";
+
+        //check if there is a transaction that completes the sandwich
+        if (callHistory.has(accountBack + token0 + token1)) {
+          let potentialSandwhichFront: [LogDescription, number, string] = callHistory.get(
+            accountBack + token0 + token1
+          )!;
+          const {
+            account: accountFront,
+            tokenIn: tokenInFront,
+            tokenOut: tokenOutFront,
+            amountIn: amountInFront,
+            amountOut: amountOutFront,
+          } = potentialSandwhichFront[0].args;
+
+          //check if the front transaction trades the same tokens in the opposite direction as the back transaction
+          if (tokenInBack == tokenOutFront && tokenOutBack == tokenInFront) {
+            //find any victim transaction
+            callHistoryKeys.forEach((addressAndTokens) => {
+              let event = callHistory.get(addressAndTokens)!;
+              const {
+                account: accountVictim,
+                tokenIn: tokenInVictim,
+                tokenOut: tokenOutVictim,
+                amountIn: amountInVictim,
+                amountOut: amountOutVictim,
+              } = event[0].args;
+              //check if victim was inside the sandwich
+              if (
+                (event[1] > potentialSandwhichFront[1] &&
+                  event[1] < callHistoryIndex &&
+                  callHistoryIndex > potentialSandwhichFront[1]) ||
+                (event[1] < potentialSandwhichFront[1] &&
+                  event[1] < callHistoryIndex &&
+                  callHistoryIndex < potentialSandwhichFront[1]) ||
+                (event[1] > potentialSandwhichFront[1] &&
+                  event[1] > callHistoryIndex &&
+                  callHistoryIndex < potentialSandwhichFront[1])
```

Comment:
I don't get what are you exactly checking in here. Can you explain ? 
I tried to understand this from a logic pov and i found out that: 

1.  **The first check:** `(event[1] > potentialSandwhichFront[1] &&  event[1] < callHistoryIndex &&  callHistoryIndex > potentialSandwhichFront[1])` is simply equivalent to 
`event[1] > potentialSandwhichFront[1] &&  event[1] < callHistoryIndex`. ie. the 3rd condition is not needed because 1+2 implies 3. 

2. **The second check:** `(event[1] < potentialSandwhichFront[1] && event[1] < callHistoryIndex && callHistoryIndex < potentialSandwhichFront[1])` is equivalent to 
`event[1] < callHistoryIndex && callHistoryIndex < potentialSandwhichFront[1]`. ie. the first condition is not needed because 2+3=> 1 . 
   
3. **The third check:**  is equivalent to `event[1] > potentialSandwhichFront[1] && callHistoryIndex < potentialSandwhichFront[1]`. ie the 2nd condition is not needed. 

Now overall it seems like you are checking if : 

```
potentialSandwhichFront[1] < event[1] < callHistoryIndex ||
event[1] < callHistoryIndex < potentialSandwhichFront[1] ||
callHistoryIndex < potentialSandwhichFront[1] < event[1]
```
I don't think this is what you intended to do, can you please explain this part.

Reply:
Some of the conditions are redundant like you pointed out, but I thought it made more sense to leave them like this for the sake of understanding. Basically this condition checks if the victim transaction is in the sandwich, but because we have an array of a limited size that is similar to a circular array, we can't just check if the index of the victim transaction is in between the front and the back one (because some transaction could find the array full and occupy an earlier index). 
So there are 3 possible situations, where in each `potentialSandwhichFront[1]` is the front transaction, `event[1]` is the victim transaction, and `callHistoryIndex` is the back transaction:

1. In the first case (assuming array size 20), we check that the victim transaction is after the front one, the victim one is before the back one, and the back one is after the front one. Some example values can be `4, 6, 9`
2. In the second case (assuming array size 20), we check that the victim transaction is before the front transaction, the victim one is before the back one, and the back one is before the front. Some example values can be `18, 10, 15`
3. In the third case (assuming array size 20), we check that the victim transaction is after the front transaction, the victim one is after the back transaction, and the back one is before the front one. Some example values can be `10, 17, 5`

To sum up: Considering a circular structure (where the element after the last one is the first one), the transactions must be in some permutation that is in ascending order